u_fingerprinter.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2017 Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// Fingerprinter is a struct largely for holding options for the FingerprintClientHello func
+type Fingerprinter struct {
+	// KeepPSK will ensure that PSK extensions are passed along into the resulting ClientHelloSpec
+	KeepPSK bool
+	// AllowBluntMimicry will ensure that unknown extensions are passed along into the resulting ClientHelloSpec
+	// WARNING: there could be numerous subtle issues with ClientHelloSpecs that are generated with this flag
+	AllowBluntMimicry bool
+}
+
+// FingerprintClientHello returns a ClientHelloSpec which is based on the ClientHello that is passed in as the data argument
+//
+// If the ClientHello passed in has extensions that are not recognized or cannot be handled, it will return a non-nil error and a nil *ClientHelloSpec value
+//
+// The data should be the ClientHello record as produced by crypto/tls.clientHelloMsg.marshal()
```

Comment:
why not the full record?
that should make the function easier to use

Reply:
I think it was a kind of least necessary information thing ie for tests (or perhaps other purposes??) where the caller only has the `ClientHello` message, it would require fabricating a the rest of the record?

I think at first I also chose it because of its parallels with `crypto/tls.clientHelloMsg.marshal()`.

I've gone ahead and changed it to accept the full record.